File: premium-chatbot/client/src/services/safety.ts
const patterns: { regex: RegExp; description: string }[] = [
  { regex: /ignore\s+(previous|all|current|your|all\s+previous)\s+(instructions|prompts|directives|messages|commands)/i, description: 'Prompt injection' },
  { regex: /reveal\s+(your\s+)?(system|hidden|internal|secret)\s+(prompt|instructions|directives|configuration|message)/i, description: 'System prompt extraction' },
  { regex: /show\s+(your\s+)?(hidden|internal|system|secret)\s+(prompt|instructions|directives|message)/i, description: 'System prompt extraction' },
  { regex: /output\s+(your\s+)?(system|hidden|internal)\s+(prompt|instructions)/i, description: 'System prompt extraction' },
  { regex: /developer\s*mode/i, description: 'Developer mode jailbreak' },
  { regex: /\bDAN\b/i, description: 'DAN jailbreak' },
  { regex: /\bjailbreak\b/i, description: 'Jailbreak attempt' },
  { regex: /print\s+(environment\s+)?(variables|secrets|keys|config|configuration)/i, description: 'Secret extraction' },
  { regex: /show\s+(api\s*key|secret|password|token|credential)/i, description: 'Credential extraction' },
  { regex: /act\s+as\s+(an?\s+)?(unrestricted|unfiltered|unlimited|free|uncensored|another)\s+AI/i, description: 'Jailbreak roleplay' },
  { regex: /you\s+(are|should)\s+(now\s+)?(free|unrestricted|unfiltered|open|no\s+(rules|limits|restrictions|guardrails))/i, description: 'Jailbreak attempt' },
  { regex: /ignore\s+(safety|security|ethics|boundaries)/i, description: 'Safety override attempt' },
  { regex: /from\s+(now\s+)?on\s*,\s*you\s+(are|will|should)/i, description: 'Identity override attempt' },
];

export interface SafetyResult {
  safe: boolean;
  reason?: string;
}

export function validatePrompt(prompt: string): SafetyResult {
  for (const { regex, description } of patterns) {
    if (regex.test(prompt)) {
      logBlocked(prompt, description);
      return { safe: false, reason: description };
    }
  }
  return { safe: true };
}

function logBlocked(prompt: string, reason: string) {
  try {
    const stored = JSON.parse(localStorage.getItem('nexaSafetyLog') || '[]');
    stored.push({
      prompt: prompt.length > 120 ? prompt.slice(0, 120) + '…' : prompt,
      reason,
      timestamp: new Date().toISOString(),
    });
    if (stored.length > 50) stored.splice(0, stored.length - 50);
    localStorage.setItem('nexaSafetyLog', JSON.stringify(stored));
  } catch {
    localStorage.removeItem('nexaSafetyLog');
  }
}

export const safetyFeatures = [
  { id: 'injection', label: 'Prompt Injection Protection' },
  { id: 'jailbreak', label: 'Jailbreak Detection' },
  { id: 'sysprompt', label: 'System Prompt Protection' },
  { id: 'secrets', label: 'Secret & Credential Protection' },
];

export const BLOCK_MESSAGE = 'I cannot process this request. It appears to attempt overriding safety guidelines or accessing protected information. Please rephrase your question.';

File: premium-chatbot/server/app/services/safety.py
import re
import os
from datetime import datetime
from typing import List, Dict, Any

PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r'reveal\s+(your\s+)?(system|hidden|internal|secret)\s+(prompt|instructions|directives|configuration)', re.I), 'System prompt extraction'),
    (re.compile(r'show\s+(your\s+)?(system|hidden|internal|secret)\s+(prompt|instructions|directives|configuration)', re.I), 'System prompt extraction'),
    (re.compile(r'print\s+(your\s+)?(system|hidden|internal|secret)\s+(prompt|instructions|directives)', re.I), 'System prompt extraction'),
    (re.compile(r'(developer|admin)\s*(instructions?|mode|override|directives)', re.I), 'Developer instructions extraction'),
    (re.compile(r'(hidden|secret|private)\s+(prompt|instructions?|message|directives)', re.I), 'Hidden prompt extraction'),
    (re.compile(r'reveal\s+(configuration|api[\s_]*key|credentials?|secrets?)', re.I), 'Configuration extraction'),
    (re.compile(r'show\s+(environment\s+variables?|env\s*vars?|api[\s_]*key|secrets?|credentials?|tokens?)', re.I), 'Environment variable extraction'),
    (re.compile(r'print\s+(secrets?|environment\s+variables?|env\s*vars?|api[\s_]*key)', re.I), 'Secret extraction'),
    (re.compile(r'\bjailbreak\b', re.I), 'Jailbreak attempt'),
    (re.compile(r'\bDAN\b', re.I), 'DAN jailbreak'),
    (re.compile(r'ignore\s+(previous|all|current|your)\s+(instructions|prompts|directives|commands)', re.I), 'Prompt injection'),
    (re.compile(r'override\s+(instructions?|prompts?|directives|rules|safety)', re.I), 'Override attempt'),
]

BLOCK_RESPONSE = "Sorry, I cannot provide internal instructions, system prompts, credentials, or protected configuration details."

LOG_FILE = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), 'safety.log')


def check_messages(messages: List[Dict[str, Any]]) -> str | None:
    """Check all user messages for safety violations. Returns block message if found, None if safe."""
    for msg in messages:
        if msg.get('role') != 'user':
            continue
        parts = msg.get('parts', [])
        for part in parts:
            text = part.get('text', '')
            for pattern, reason in PATTERNS:
                if pattern.search(text):
                    _log_block(text, reason)
                    return BLOCK_RESPONSE
    return None


def _log_block(prompt: str, reason: str) -> None:
    """Log blocked attempt to file."""
    try:
        timestamp = datetime.now().isoformat()
        truncated = prompt[:150] + '…' if len(prompt) > 150 else prompt
        with open(LOG_FILE, 'a', encoding='utf-8') as f:
            f.write(f'[{timestamp}] BLOCKED | {reason} | "{truncated}"\n')
    except Exception:
        pass
